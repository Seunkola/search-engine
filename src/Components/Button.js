import React, { Component } from 'react';

class Button extends Component {
    render(){
        const {
        onClick,
        className,
        children
        } = this.props;

        return(
            <button 
                className={className} 
                onClick={onClick}
                type="button"
                >
                {
                    !children
                    ?
                    <i className="fas fa-trash-alt"></i>
                    :
                    children
                }
            </button>
        );
    }
}

Button.defaultProps = {
    className: '',
};

export default Button;