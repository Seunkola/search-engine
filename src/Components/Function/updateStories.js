export const updateSearchTopStoriesState = (hits, page) => (prevState) => {
    const { searchKey, results } = prevState;

    const oldHits = results && results[searchKey]
    ? results[searchKey].hits
    : [];

    const updatedHits = [
    ...oldHits,
    ...hits
    ];

    return {
        results: {
        ...results,
        [searchKey]: { hits: updatedHits, page }
        },
        isLoading: false
    };
};
    