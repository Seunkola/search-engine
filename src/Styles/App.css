.App {
  text-align: center;
}

.App-logo {
  animation: App-logo-spin infinite 20s linear;
  height: 40vmin;
  pointer-events: none;
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #61dafb;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}

.page {
  margin: 20px;
  }

.interactions {
  text-align: center;
}

.theList {
  margin: 20px 0;
}

.theList-header {
  display: flex;
  line-height: 24px;
  white-space: nowrap;
  margin: 10px 0;
  padding: 10px;
  background: blue;
  border: 1px solid #e3e3e3;
}

.theList-empty {
  margin: 200px;
  text-align: center;
  font-size: 16px;
}

.theList-row {
  display: flex;
  line-height: 24px;
  white-space: nowrap;
  margin: 10px 0;
  padding: 10px;
  background: #ffffff;
  border: 1px solid #e3e3e3;
}

.theList-header > span {
  overflow: hidden;
  text-overflow: ellipsis;
  padding: 0 5px;
  color: white;
}

.theList-row > span {
  overflow: hidden;
  text-overflow: ellipsis;
  padding: 0 5px;
}

.button-inline {
  border-width: 0;
  background: transparent;
  color: inherit;
  text-align: inherit;
  -webkit-font-smoothing: inherit;
  padding: 0;
  font-size: inherit;
  cursor: pointer;
}

.button-active {
  border-radius: 0;
  border-bottom: 1px solid #38BB6C;
}

.theList-header > span > i {
  margin-left: 2%;
}