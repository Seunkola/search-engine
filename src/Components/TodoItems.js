import React, { Component } from 'react'
import Button from './Button';
import { SORTS } from '../Constants/Constants';
import ListHeader from './TodoItemsHeader';

const largeColumn = {
  width: '20%',
};

const midColumn = {
  width: '5%',
};
  
class TodoItems extends Component {
  render() {
    const {
      sortKey,
      onSort,
      isSortReverse,
    } = this.props;

    const todoEntries = this.props.entries;
    const sortedList = SORTS[sortKey](todoEntries);
    const reverseSortedList = isSortReverse ? sortedList.reverse() : sortedList;

    return (
      <div className="theList">
        <ListHeader onSort={onSort} />
        {reverseSortedList.map(item =>
          <div key={item.objectID} className="theList-row">
            <span style={largeColumn}>
              {item.title}
            </span>
            <span style={largeColumn}>
              {item.author}
            </span>
            <span style={largeColumn}>
              {item.num_comments}
            </span>
            <span style={largeColumn}>
              {item.points}
            </span>
            <span style={midColumn}>
              <a href={item.url}>View</a>
            </span>
            <span style={midColumn}>
              <Button
                className="button-inline"
                onClick={() => this.props.onDismiss(item.objectID)}
                >
              </Button>
            </span>
          </div>
        )}
      </div>
      );
  }
}

export default TodoItems;