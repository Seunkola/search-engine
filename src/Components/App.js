import React, { Component } from 'react';
import '../Styles/App.css';
import TodoItems  from './TodoItems';
import Header from './TodoList';
import { DEFAULT_QUERY, PATH_BASE, PATH_SEARCH, PARAM_SEARCH, PARAM_PAGE, PARAM_HPP, DEFAULT_HPP,
} from '../Constants/Constants';
import Button from './Button';
import ErrorDisplay from './ErrorDisplay';
import axios from 'axios';
import PropTypes from 'prop-types';
import Loading from './Loading';
import { updateSearchTopStoriesState } from './Function/updateStories';

class App extends Component {
  constructor(props) {
    super(props);

    this.state = {
      results: null,
      searchKey: '',
      searchTerm: DEFAULT_QUERY,
      error: null,
      isLoading: false,
      sortKey: 'NONE',
      isSortReverse: false,
    };

    this.needToSearchTopStories = this.needToSearchTopStories.bind(this);
    this.setSearchTopStories = this.setSearchTopStories.bind(this);
    this.onSearchChange = this.onSearchChange.bind(this);
    this.onSearchSubmit = this.onSearchSubmit.bind(this);
    this.onDismiss = this.onDismiss.bind(this);
    this.onSort = this.onSort.bind(this);
  }
  
  needToSearchTopStories(searchTerm) {
    return !this.state.results[searchTerm];
  }

  setSearchTopStories(result) {
    const { hits, page } = result;
    this.setState(updateSearchTopStoriesState(hits,page));
  }

  fetchSearchTopStories(searchTerm, page = 0) {
    this.setState({ isLoading: true });

    axios(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`)
      .then(result => this.setSearchTopStories(result.data))
      .catch(error => 
        this.setState({ error })
      );
  }

  componentDidMount() {
    const { searchTerm } = this.state;
    this.setState({ searchKey: searchTerm });
    this.fetchSearchTopStories(searchTerm);
  }

  onSearchChange(event) {
    this.setState({ searchTerm: event.target.value });
  }

  onSearchSubmit(event) {
    const { searchTerm } = this.state;
    this.setState({ searchKey: searchTerm });
    if(this.needToSearchTopStories(searchTerm)){
      this.fetchSearchTopStories(searchTerm);
    }
    
    event.preventDefault();
  }

  onSort(sortKey) {
    const isSortReverse = this.state.sortKey === sortKey && !this.state.isSortReverse;
    this.setState({ sortKey, isSortReverse });
  }

  onDismiss(id) {
    const { searchKey, results } = this.state;
    const { hits, page } = results[searchKey];

    const isNotId = item => item.objectID !== id;
    const updatedHits = hits.filter(isNotId);
    this.setState({ 
      results: { 
        ...results, 
        [searchKey]: {hits: updatedHits, page}
       }
     });
  }

  render(){
    const { searchTerm, results, searchKey, error, isLoading, sortKey, isSortReverse } = this.state;
    const page = (results && results[searchKey] && results[searchKey].page) || 0;    
    const list = (results && results[searchKey] && results[searchKey].hits) || [];

    return (
      <div className="App">
        <Header 
          onSearchChange={this.onSearchChange} 
          searchTodo={searchTerm}
          onSearchSubmit={this.onSearchSubmit} />

        <ErrorDisplay error={error} />

        <TodoItems 
        entries={list}
        sortKey={sortKey}
        onSort={this.onSort} 
        onDismiss={this.onDismiss}
        isSortReverse={isSortReverse} 
        />

        <div className="interactions">
        {
          isLoading 
          ?
          <Loading />
          :
          <Button onClick={() => this.fetchSearchTopStories(searchKey, page + 1)}>
            Load More...
          </Button>
        }
        </div>
      </div>
    );
  }
}

Button.PropTypes = {
  onClick: PropTypes.func.isRequired,
  className: PropTypes.string,
  children: PropTypes.node.isRequired,
}

TodoItems.PropTypes = {
  entries: PropTypes.arrayOf(
    PropTypes.shape({
      objectID: PropTypes.string.isRequired,
      title: PropTypes.string,
      author: PropTypes.string,
      url: PropTypes.string,
      num_comments: PropTypes.number,
      points: PropTypes.number,
    })
  )
  .isRequired,
  onDismiss: PropTypes.func.isRequired,
}

Header.PropTypes = {
  searchTodo: PropTypes.string.isRequired,
  onSearchSubmit: PropTypes.func.isRequired,
  onSearchChange: PropTypes.func,
}

export default App;