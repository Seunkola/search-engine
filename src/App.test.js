import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import App from './Components/App';
import Header from './Components/TodoList';
import Button from './Components/Button';
import TodoItems from './Components/TodoItems';
import { updateSearchTopStoriesState } from './Components/Function/updateStories';

Enzyme.configure({ adapter: new Adapter() });

describe('App', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create( <App /> );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});

describe('Header with search', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Header />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create( <Header /> );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});

describe('Button', () => {

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Button  onClick={() => 'test'}>Test</Button>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create( <Button onClick={() => 'test'}>Test</Button> );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});

describe('Display List of Items', () => {
  const props = {
    list: [
      { title: '1', objectID: '1'},
      { title: '2', objectID: '2'},
    ],
    sortKey: 'TITLE',
    isSortReverse: false,
  };

  it('renders without crashing', () => {
    const div =  document.createElement('div');
    ReactDOM.render(<TodoItems 
      entries={props.list}
        sortKey={props.sortKey}
        isSortReverse={props.isSortReverse} />, div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create( 
    <TodoItems  
      entries={props.list}
      sortKey={props.sortKey}
      isSortReverse={props.isSortReverse} 
      /> );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('shows two items in the List', () => {
    const element = shallow( 
    <TodoItems 
    entries={props.list}
    sortKey={props.sortKey}
    isSortReverse={props.isSortReverse}
     /> );
    expect(element.find('.theList-row').length).toBe(2);
  });

});

describe('updates searched stories', () => {
  it('returns a new state', () => {

    const results = {
      hits: [
        { title: '1', objectID: '1'},
        { title: '2', objectID: '2'},
      ],
      page: 1,
    }

    const { hits, page } = results;

    const wrapper = shallow(<App />);

    wrapper.setState(updateSearchTopStoriesState(hits, page));

    const expectedStateResults = {
        "redux": {
            "hits": [
              { title: '1', objectID: '1'},
              { title: '2', objectID: '2'},
            ],
          "page": 1
        },
    };

    expect(wrapper.state('results')).toEqual(expectedStateResults);
  
  });
});

