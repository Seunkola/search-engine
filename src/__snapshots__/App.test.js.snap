// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App has a valid snapshot 1`] = `
<div
  className="App"
>
  <div
    className="todoListMain"
  >
    <div
      className="Header"
    >
      <form
        onSubmit={[Function]}
      >
        <input
          id="inputBox"
          onChange={[Function]}
          placeholder="Topic"
          value="redux"
        />
        <button
          id="addTask"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
  </div>
  <div
    className="theList"
  />
  <div
    className="theList"
  >
    <div
      className="theList-header"
    >
      <span
        style={
          Object {
            "width": "20%",
          }
        }
      >
        <button
          className="button-inline"
          onClick={[Function]}
          type="button"
        >
          Title
        </button>
        <i
          className="fas fa-sort"
        />
      </span>
      <span
        style={
          Object {
            "width": "20%",
          }
        }
      >
        <button
          className="button-inline button-active"
          onClick={[Function]}
          type="button"
        >
          Author
        </button>
        <i
          className="fas fa-sort"
        />
      </span>
      <span
        style={
          Object {
            "width": "20%",
          }
        }
      >
        <button
          className="button-inline button-active"
          onClick={[Function]}
          type="button"
        >
          Comments
        </button>
        <i
          className="fas fa-sort"
        />
      </span>
      <span
        style={
          Object {
            "width": "20%",
          }
        }
      >
        <button
          className="button-inline button-active"
          onClick={[Function]}
          type="button"
        >
          Points
        </button>
        <i
          className="fas fa-sort"
        />
      </span>
      <span
        style={
          Object {
            "width": "5%",
          }
        }
      >
        Read
      </span>
      <span
        style={
          Object {
            "width": "5%",
          }
        }
      >
        Remove
      </span>
    </div>
  </div>
  <div
    className="interactions"
  >
    <div>
      <div
        className="lds-ellipsis"
      >
        <div />
        <div />
        <div />
        <div />
      </div>
      Loading...
    </div>
  </div>
</div>
`;

exports[`Display List of Items has a valid snapshot 1`] = `
<div
  className="theList"
>
  <div
    className="theList-header"
  >
    <span
      style={
        Object {
          "width": "20%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Title
      </button>
      <i
        className="fas fa-sort"
      />
    </span>
    <span
      style={
        Object {
          "width": "20%",
        }
      }
    >
      <button
        className="button-inline button-active"
        onClick={[Function]}
        type="button"
      >
        Author
      </button>
      <i
        className="fas fa-sort"
      />
    </span>
    <span
      style={
        Object {
          "width": "20%",
        }
      }
    >
      <button
        className="button-inline button-active"
        onClick={[Function]}
        type="button"
      >
        Comments
      </button>
      <i
        className="fas fa-sort"
      />
    </span>
    <span
      style={
        Object {
          "width": "20%",
        }
      }
    >
      <button
        className="button-inline button-active"
        onClick={[Function]}
        type="button"
      >
        Points
      </button>
      <i
        className="fas fa-sort"
      />
    </span>
    <span
      style={
        Object {
          "width": "5%",
        }
      }
    >
      Read
    </span>
    <span
      style={
        Object {
          "width": "5%",
        }
      }
    >
      Remove
    </span>
  </div>
  <div
    className="theList-row"
  >
    <span
      style={
        Object {
          "width": "20%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "20%",
        }
      }
    />
    <span
      style={
        Object {
          "width": "20%",
        }
      }
    />
    <span
      style={
        Object {
          "width": "20%",
        }
      }
    />
    <span
      style={
        Object {
          "width": "5%",
        }
      }
    >
      <a>
        View
      </a>
    </span>
    <span
      style={
        Object {
          "width": "5%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        <i
          className="fas fa-trash-alt"
        />
      </button>
    </span>
  </div>
  <div
    className="theList-row"
  >
    <span
      style={
        Object {
          "width": "20%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "20%",
        }
      }
    />
    <span
      style={
        Object {
          "width": "20%",
        }
      }
    />
    <span
      style={
        Object {
          "width": "20%",
        }
      }
    />
    <span
      style={
        Object {
          "width": "5%",
        }
      }
    >
      <a>
        View
      </a>
    </span>
    <span
      style={
        Object {
          "width": "5%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        <i
          className="fas fa-trash-alt"
        />
      </button>
    </span>
  </div>
</div>
`;

exports[`Header with search has a valid snapshot 1`] = `
<div
  className="todoListMain"
>
  <div
    className="Header"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        id="inputBox"
        onChange={[Function]}
        placeholder="Topic"
        value="test"
      />
      <button
        id="addTask"
        type="submit"
      >
        Search
      </button>
    </form>
  </div>
</div>
`;
